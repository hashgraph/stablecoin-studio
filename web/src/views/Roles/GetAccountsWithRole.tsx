import { Heading, useDisclosure } from '@chakra-ui/react';
import { GetAccountsWithRolesRequest, StableCoinRole } from '@hashgraph/stablecoin-npm-sdk';
import { SelectController } from '../../components/Form/SelectController';
import { useState } from 'react';
import { useFieldArray, useForm } from 'react-hook-form';
import { useTranslation } from 'react-i18next';
import { useSelector } from 'react-redux';

import type { ModalsHandlerActionsProps } from '../../components/ModalsHandler';
import ModalsHandler from '../../components/ModalsHandler';

import { SDKService } from '../../services/SDKService';
import { SELECTED_NETWORK, SELECTED_WALLET_COIN } from '../../store/slices/walletSlice';

import OperationLayout from '../Operations/OperationLayout';
import DetailsReview from '../../components/DetailsReview';
import type { Detail } from '../../components/DetailsReview';

const GetAccountsWithRole = () => {
	const { t } = useTranslation(['global', 'roles', 'stableCoinCreation', 'externalTokenInfo']);
	const {
		isOpen: isOpenModalAction,
		onOpen: onOpenModalAction,
		onClose: onCloseModalAction,
	} = useDisclosure();
	const selectedStableCoin = useSelector(SELECTED_WALLET_COIN);

	const { control, getValues, setError } = useForm({
		mode: 'onChange',
	});

	const network = useSelector(SELECTED_NETWORK);
	const hashScanURL = `https://hashscan.io/${network}`;

	useFieldArray({
		control,
		name: 'rol',
	});

	const [errorTransactionUrl, setErrorTransactionUrl] = useState();

	const [accountsWithRoles, setAccountsWithRoles] = useState<String[]>([]);

	const details: Detail[] = [];
	accountsWithRoles &&
		accountsWithRoles.forEach((accountWithRole) => {
			details.push({
				label: t('account'),
				value: accountWithRole,
				copyButton: true,
				hashScanURL: `${hashScanURL}/account/${accountWithRole}`,
			});
		});

	const supplyTypes = [
		{
			value: StableCoinRole.DEFAULT_ADMIN_ROLE,
			label: t('roles:getAccountsWithRole.options.Admin'),
		},
		{
			value: StableCoinRole.CASHIN_ROLE,
			label: t('roles:getAccountsWithRole.options.CashIn'),
		},
		{
			value: StableCoinRole.FREEZE_ROLE,
			label: t('roles:getAccountsWithRole.options.Freeze'),
		},
		{
			value: StableCoinRole.PAUSE_ROLE,
			label: t('roles:getAccountsWithRole.options.Pause'),
		},
		{
			value: StableCoinRole.BURN_ROLE,
			label: t('roles:getAccountsWithRole.options.Burn'),
		},
		{
			value: StableCoinRole.RESCUE_ROLE,
			label: t('roles:getAccountsWithRole.options.Rescue'),
		},
		{
			value: StableCoinRole.WIPE_ROLE,
			label: t('roles:getAccountsWithRole.options.Wipe'),
		},
		{
			value: StableCoinRole.DELETE_ROLE,
			label: t('roles:getAccountsWithRole.options.Delete'),
		},
		{
			value: StableCoinRole.KYC_ROLE,
			label: t('roles:getAccountsWithRole.options.Kyc'),
		},
	];
	const selectorStyle = {
		wrapper: {
			border: '1px',
			borderColor: 'brand.black',
			borderRadius: '8px',
			height: 'min',
		},
		menuList: {
			maxH: '220px',
			overflowY: 'auto',
			bg: 'brand.white',
			boxShadow: 'down-black',
			p: 2,
			zIndex: 99,
		},
		valueSelected: {
			fontSize: '14px',
			fontWeight: '500',
		},
	};

	const handleGetAccountsWithRole: ModalsHandlerActionsProps['onConfirm'] = async ({
		onSuccess,
		onError,
		onLoading,
	}) => {
		onLoading();
		const values = getValues();

		console.log(values);
		const request = new GetAccountsWithRolesRequest({
			roleId: values.supplyType.value,
			tokenId: selectedStableCoin!.tokenId!.toString(),
		});

		console.log(request);
		try {
			const response: any = await Promise.race([
				SDKService.getAccountsWithRole(request),
				new Promise((resolve, reject) => {
					setTimeout(() => {
						reject(new Error("Account's roles couldn't be obtained in a reasonable time."));
					}, 10000);
				}),
			]).catch((e) => {
				console.log(e.message);
				onOpenModalAction();
				throw e;
			});
			setAccountsWithRoles(response);
			onSuccess();
		} catch (error: any) {
			console.log(error);
			// is MultiTargetsInvalid

			setErrorTransactionUrl(error.transactionUrl);
			onError();
		}
	};

	const handleSubmit = () => {
		const values = getValues().rol;

		const valuesDuplicated: { [index: string]: number[] } = {};
		let sendRequest = true;
		values.forEach((obj: any, index: number) => {
			if (!valuesDuplicated[obj.accountId]) {
				valuesDuplicated[obj.accountId] = [];
			}
			valuesDuplicated[obj.accountId].push(index);
		});
		for (const accountId in valuesDuplicated) {
			if (valuesDuplicated[accountId].length > 1) {
				sendRequest = false;
				valuesDuplicated[accountId].map((index: number) =>
					setError(`rol[${index}].accountId`, {
						type: 'repeatedValue',
						message: t('roles:giveRole.errorAccountIdDuplicated', {
							account: accountId,
						})!,
					}),
				);
			}
		}

		if (sendRequest) {
			onOpenModalAction();
		}
	};

	return (
		<>
			<OperationLayout
				LeftContent={
					<>
						<Heading data-testid='title' fontSize='24px' fontWeight='700' mb={10} lineHeight='16px'>
							{t(`roles:getAccountsWithRole.title`, { role: '' })}
						</Heading>

						<SelectController
							control={control}
							name={'supplyType'}
							options={supplyTypes}
							label={t('roles:getAccountsWithRole.chooseRole')}
							placeholder={t('stableCoinCreation:optionalDetails.typeSupplyPlaceholder')}
							overrideStyles={selectorStyle}
							addonLeft={true}
							variant='unstyled'
							defaultValue={'0'}
						/>

						<DetailsReview
							title={t(`roles:getAccountsWithRole.subtitle`, {
								role: getValues()?.supplyType?.label,
							})}
							titleProps={{ fontWeight: 'bold' }}
							contentProps={{ justifyContent: 'space-between', gap: 4 }}
							details={details}
						/>
					</>
				}
				onConfirm={handleSubmit}
				confirmBtnProps={{
					isDisabled: false,
				}}
			/>

			<ModalsHandler
				errorNotificationTitle={t(`roles:getAccountsWithRole.modal.title`)}
				// @ts-ignore-next-line
				errorNotificationDescription={t(`roles:revokeRole.modalErrorDescription`)}
				errorTransactionUrl={errorTransactionUrl}
				// @ts-ignore-next-line
				warningNotificationDescription={t(`roles:revokeRole.modalErrorDescription`)}
				modalActionProps={{
					isOpen: isOpenModalAction,
					onClose: onCloseModalAction,
					title: t(`roles:getAccountsWithRole.modal.title`) + ' ' + getValues()?.supplyType?.label,
					confirmButtonLabel: t(`roles:getAccountsWithRole.modalActionConfirmButton`),
					onConfirm: handleGetAccountsWithRole,
				}}
				ModalActionChildren={<></>}
				successNotificationTitle={
					t(`roles:getAccountsWithRole.modal.title`) + ' ' + getValues()?.supplyType?.label
				}
			/>
		</>
	);
};

export default GetAccountsWithRole;
