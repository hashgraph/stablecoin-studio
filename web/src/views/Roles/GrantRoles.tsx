import {
	Text,
	Heading,
	CheckboxGroup,
	Grid,
	useDisclosure,
	Flex,
	Button,
	Box,
	HStack,
} from '@chakra-ui/react';
import {
	GrantRoleRequest,
	GrantMultiRolesRequest,
	StableCoinRole,
	GetRolesRequest,
} from '@hashgraph/stablecoin-npm-sdk';
import React, { useEffect, useMemo, useState } from 'react';
import { useFieldArray, useForm } from 'react-hook-form';
import { useTranslation } from 'react-i18next';
import { useDispatch, useSelector } from 'react-redux';
import type { Detail } from '../../components/DetailsReview';
import DetailsReview from '../../components/DetailsReview';
import { CheckboxController } from '../../components/Form/CheckboxController';
import InputController from '../../components/Form/InputController';
import SwitchController from '../../components/Form/SwitchController';
import Icon from '../../components/Icon';
import type { ModalsHandlerActionsProps } from '../../components/ModalsHandler';
import ModalsHandler from '../../components/ModalsHandler';
import { propertyNotFound } from '../../constant';
import { SDKService } from '../../services/SDKService';
import {
	SELECTED_WALLET_COIN,
	SELECTED_WALLET_PAIRED_ACCOUNTID,
	walletActions,
} from '../../store/slices/walletSlice';
import type { AppDispatch } from '../../store/store';
import { handleRequestValidation, validateDecimalsString } from '../../utils/validationsHelper';
import OperationLayout from '../Operations/OperationLayout';

interface GrantRoleForm {
	accountId: string;
	amount?: string;
	infinity?: boolean;
}

const GrantRoleOperation = ({
	filteredCapabilities,
}: {
	filteredCapabilities: { id: string; value: StableCoinRole; label: string }[];
}) => {
	const dispatch = useDispatch<AppDispatch>();
	const { t } = useTranslation(['global', 'roles', 'stableCoinCreation', 'externalTokenInfo']);
	const {
		isOpen: isOpenModalAction,
		onOpen: onOpenModalAction,
		onClose: onCloseModalAction,
	} = useDisclosure();
	const selectedStableCoin = useSelector(SELECTED_WALLET_COIN);
	const accountId = useSelector(SELECTED_WALLET_PAIRED_ACCOUNTID);
	const { control, watch, getValues, formState, setError } = useForm({
		mode: 'onChange',
	});
	const {
		fields: accounts,
		append,
		remove,
	} = useFieldArray({
		control,
		name: 'rol',
	});
	const isMaxAccounts = useMemo(() => accounts.length >= 10, [accounts]);
	const isRoleSelected = useMemo((): boolean => {
		const values = watch();
		delete values.rol;
		if (!values) return false;
		return Object.values(values).some((item) => {
			return item === true;
		});
	}, [watch()]);
	const [errorTransactionUrl, setErrorTransactionUrl] = useState();
	const [grantRoles, setGrantRoles] = useState<GrantRoleRequest[]>([]);

	useEffect(() => {
		addNewAccount();
	}, []);

	const handleGrantRoles: ModalsHandlerActionsProps['onConfirm'] = async ({
		onSuccess,
		onError,
		onLoading,
		onCloseModalLoading,
	}) => {
		onLoading();
		const values = getValues();
		const rolesRequest: string[] = [];
		for (const key in values) {
			if (values[key] === true) {
				rolesRequest.push(filteredCapabilities.find((item) => item.id === key)!.value);
			}
		}
		const targets = values.rol.map((item: GrantRoleForm) => item.accountId);
		const request = new GrantMultiRolesRequest({
			tokenId: selectedStableCoin!.tokenId!.toString(),
			targetsId: targets,
			roles: rolesRequest as StableCoinRole[],
		});
		if (rolesRequest.includes(StableCoinRole.CASHIN_ROLE)) {
			request.amounts = values.rol.map((item: GrantRoleForm) =>
				item.infinity ? '0' : item.amount,
			);
		}

		try {
			await SDKService.grantMultipleRole(request);
			// Update current account role if is target
			const isAccountTarget = targets.some(
				(item: any) => item?.toString() === accountId?.toString(),
			);
			if (isAccountTarget) {
				const roles = await Promise.race([
					SDKService.getRoles(
						new GetRolesRequest({
							tokenId: selectedStableCoin!.tokenId!.toString(),
							targetId: accountId!.toString(),
						}),
					),
					new Promise((resolve, reject) => {
						setTimeout(() => {
							reject(new Error("Account's roles couldn't be obtained in a reasonable time."));
						}, 10000);
					}),
				]).catch((e) => {
					console.log(e.message);
					onOpenModalAction();
					throw e;
				});
				dispatch(walletActions.setRoles(roles));
			}
			onSuccess();
		} catch (error: any) {
			// is MultiTargetsInvalid
			if ('targetsId' in error) {
				const targetsNoExists = error.targetsId;
				values.rol.forEach((value: any, index: number) => {
					if (targetsNoExists.find((item: any) => item === value.accountId)) {
						setError(`rol[${index}].accountId`, {
							type: 'invalidAccount',
							message: t('roles:giveRole.errorAccountIdNotExists', {
								account: value.accountId,
							})!,
						});
					}
				});
				onCloseModalLoading();
			} else {
				setErrorTransactionUrl(error.transactionUrl);
				onError();
			}
		}
	};
	const getAccountsDetails = () => {
		const values = getValues();

		if (values.rol) {
			const details: Detail[] = values.rol.map((item: GrantRoleForm) => {
				return {
					label: t(`roles:giveRole.modalActionDetailAccount`),
					value: item.accountId,
				};
			});
			return details;
		}

		return [] as Detail[];
	};
	const getRolesDetails = () => {
		const obj = getValues();
		const asArray = Object.entries(obj);

		// eslint-disable-next-line @typescript-eslint/no-unused-vars
		const filtered = asArray.filter(([key, value]) => value === true);

		return filtered.map((item) => {
			return {
				label: t(`roles:giveRole.modalActionDetailRole`),
				value: item[0],
			};
		});
	};
	const addNewAccount = () => {
		if (accounts.length >= 10) return;
		const currentGrantRole = grantRoles;
		currentGrantRole.push(
			new GrantRoleRequest({
				tokenId: selectedStableCoin!.tokenId!.toString(),
				targetId: '',
				role: undefined,
				amount: '0',
			}),
		);
		setGrantRoles(currentGrantRole);
		append({ accountId: '', amount: '', infinity: true });
	};
	const removeAccount = (i: number) => {
		if (accounts.length === 1) return;
		const currentRoleRequest = grantRoles;
		currentRoleRequest.splice(i, 1);
		setGrantRoles(currentRoleRequest);
		remove(i);
	};

	const renderSupplierQuantity = (index: number) => {
		const { decimals = 0 } = selectedStableCoin || {};
		return (
			<>
				<Box>
					<HStack ml='16px' mt='24px' width='216px'>
						<Text mr='10px'>{t(`roles:giveRole.switchLabel`)}</Text>
						<SwitchController control={control} name={`rol.${index}.infinity` as const} />
					</HStack>
				</Box>
				{!watch(`rol.${index}.infinity`) && (
					<Box ml='16px'>
						<InputController
							data-testid='input-supplier-quantity'
							rules={{
								required: t(`global:validations.required`) ?? propertyNotFound,
								validate: {
									validDecimals: (value: string) => {
										return (
											validateDecimalsString(value, decimals) ||
											(t('global:validations.decimalsValidation') ?? propertyNotFound)
										);
									},
									validation: (value: string) => {
										const _grantRole = grantRoles[index];
										_grantRole.amount = value;
										const res = handleRequestValidation(_grantRole.validate('amount'));
										return res;
									},
								},
							}}
							label={t(`roles:giveRole.amountLabel`).toString()}
							isRequired
							control={control}
							name={`rol.${index}.amount` as const}
							placeholder={t(`roles:giveRole.supplierQuantityInputPlaceholder`) ?? propertyNotFound}
						/>
					</Box>
				)}
			</>
		);
	};

	const isNotValidAccount = () => {
		const values = getValues()?.rol;
		return values
			? values.some((item: any) => {
					return item.accountId === '';
			  })
			: false;
	};

	const handleSubmit = () => {
		const values = getValues().rol;

		const valuesDuplicated: { [index: string]: number[] } = {};
		let sendRequest = true;
		values.forEach((obj: any, index: number) => {
			if (!valuesDuplicated[obj.accountId]) {
				valuesDuplicated[obj.accountId] = [];
			}
			valuesDuplicated[obj.accountId].push(index);
		});
		for (const accountId in valuesDuplicated) {
			if (valuesDuplicated[accountId].length > 1) {
				sendRequest = false;
				valuesDuplicated[accountId].map((index: number) =>
					setError(`rol[${index}].accountId`, {
						type: 'repeatedValue',
						message: t('roles:giveRole.errorAccountIdDuplicated', {
							account: accountId,
						})!,
					}),
				);
			}
		}

		if (sendRequest) {
			onOpenModalAction();
		}
	};

	return (
		<>
			<OperationLayout
				LeftContent={
					<>
						<Heading data-testid='title' fontSize='24px' fontWeight='700' mb={10} lineHeight='16px'>
							{t('roles:giveRole.title')}
						</Heading>
						<Text color='brand.gray' fontSize='24px'>
							{t(`roles:giveRole.titleRoleSection`)}
						</Text>
						<CheckboxGroup>
							<Grid column='4' gap={{ base: 4 }} templateColumns='repeat(4, 1fr)'>
								{filteredCapabilities.map((item, index) => {
									return (
										<CheckboxController
											key={index}
											value={item.value}
											control={control}
											id={`${item.label?.toString().toLocaleLowerCase()}`}
										>
											{item.label}
										</CheckboxController>
									);
								})}
							</Grid>
						</CheckboxGroup>
						<Text color='brand.gray' fontSize='24px'>
							{t(`roles:giveRole.titleAccountSection`)}
						</Text>
						{accounts &&
							accounts.map((item, i) => {
								return (
									<React.Fragment key={item.id}>
										<Flex alignItems='flex-start'>
											<Flex>
												<InputController
													key={i}
													rules={{
														required: t('global:validations.required') ?? propertyNotFound,
														validate: {
															validation: (value: string) => {
																const _grantRole = grantRoles[i];
																_grantRole.targetId = value;
																const res = handleRequestValidation(
																	_grantRole.validate('targetId'),
																);
																return res;
															},
														},
													}}
													isRequired
													formStyle={{
														width: '216px',
													}}
													control={control}
													name={`rol.${i}.accountId` as const}
													label={t(`roles:giveRole.accountLabel`).toString()}
													placeholder={t(`roles:giveRole.accountPlaceholder`).toString()}
												/>
											</Flex>
											{watch('cash in') && renderSupplierQuantity(i)}
											<Icon
												name='Trash'
												color='brand.primary'
												cursor='pointer'
												fontSize='22px'
												onClick={() => removeAccount(i)}
												// alignSelf='flex-end'
												marginLeft={{ base: 4 }}
												// marginBottom='10px'
												marginTop='40px'
											/>
										</Flex>
									</React.Fragment>
								);
							})}

						<Flex justify='flex-end' pt={6} pb={6} justifyContent='space-between'>
							<Button variant='primary' onClick={addNewAccount} isDisabled={isMaxAccounts}>
								{t(`roles:revokeRole.buttonAddAccount`)}
							</Button>
						</Flex>
					</>
				}
				onConfirm={handleSubmit}
				confirmBtnProps={{
					isDisabled: isNotValidAccount() || !isRoleSelected || !formState.isValid,
				}}
			/>

			<ModalsHandler
				errorNotificationTitle={
					errorTransactionUrl
						? t(`roles:giveRole.modalErrorTitle`)
						: t(`roles:getRole.modalErrorTitle`)
				}
				// @ts-ignore-next-line
				errorNotificationDescription={
					errorTransactionUrl
						? t(`roles:giveRole.modalErrorDescription`)
						: t(`roles:getRole.modalErrorDescription`)
				}
				errorTransactionUrl={errorTransactionUrl}
				// @ts-ignore-next-line
				warningNotificationDescription={
					errorTransactionUrl
						? t(`roles:giveRole.modalErrorDescription`)
						: t(`roles:getRole.modalErrorDescription`)
				}
				modalActionProps={{
					isOpen: isOpenModalAction,
					onClose: onCloseModalAction,
					title: t(`roles:giveRole.modalActionTitle`),
					confirmButtonLabel: t(`roles:giveRole.modalActionConfirmButton`),
					onConfirm: handleGrantRoles,
				}}
				ModalActionChildren={
					<>
						<DetailsReview
							title={t(`roles:giveRole.modalActionSubtitleAccountSection`)}
							details={getAccountsDetails()}
							divider={true}
						/>
						<DetailsReview
							title={t(`roles:giveRole.modalActionSubtitleRolesSection`)}
							details={getRolesDetails()}
						/>
					</>
				}
				successNotificationTitle={t(`roles:giveRole.modalSuccessTitle`)}
				successNotificationDescription={
					// eslint-disable-next-line no-constant-condition
					false // checkOptionSelected
						? t(`roles:giveRole.checkCashinLimitSuccessDesc`, {
								account: 'a',
								limit: 'b',
						  })
						: ''
				}
			/>
		</>
	);
};

export default GrantRoleOperation;
