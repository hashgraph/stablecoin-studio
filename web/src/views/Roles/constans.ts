export const fakeOptions = [
	{ value: 1, label: 'Admin' },
	{ value: 2, label: 'Supplier' },
	{ value: 3, label: 'Other' },
];

export const fields = {
	amount: 'amount',
	account: 'account',
	role: 'role',
	supplierQuantitySwitch: 'supplierQuantitySwitch',
};
