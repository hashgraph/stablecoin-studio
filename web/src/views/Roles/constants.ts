import { StableCoinRole } from 'hedera-stable-coin-sdk';
import type { Action } from './HandleRoles';

export const actions: Record<string, Action> = {
	edit: 'editRole' as Action,
	give: 'giveRole' as Action,
	revoke: 'revokeRole' as Action,
};

export const roleOptions = [
	{ value: StableCoinRole.CASHIN_ROLE, label: 'Cash in' },
	{ value: StableCoinRole.BURN_ROLE, label: 'Burn' },
	{ value: StableCoinRole.WIPE_ROLE, label: 'Wipe' },
	{ value: StableCoinRole.RESCUE_ROLE, label: 'Rescue' },
	{ value: StableCoinRole.PAUSER_ROLE, label: 'Pause' },
];

export const cashinLimitOptions = [
	{ value: 'INCREASE', label: 'Increase cash in limit' },
	{ value: 'DECREASE', label: 'Decrease cash in limit' },
	{ value: 'RESET', label: 'Reset cash in limit' },
	{ value: 'CHECK', label: 'Check cash in limit' },
];

export const fields = {
	amount: 'amount',
	account: 'account',
	role: 'role',
	cashinLimitOption: 'cashinLimitOption',
	name: 'name',
	supplierQuantitySwitch: 'supplierQuantitySwitch',
};
