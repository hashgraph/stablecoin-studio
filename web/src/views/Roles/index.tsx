import { Box, Heading } from '@chakra-ui/react';
import { useTranslation } from 'react-i18next';
import BaseContainer from '../../components/BaseContainer';
import { NamedRoutes } from '../../Router/NamedRoutes';
import GridDirectAccess from '../../components/GridDirectAccess';

const Roles = () => {
	const { t } = useTranslation('roles');
	const directAccesses = [
		{
			icon: 'PlusCircle',
			route: NamedRoutes.GiveRole,
			title: t('give'),
		},
		{
			icon: 'MinusCircle',
			route: NamedRoutes.RevokeRole,
			title: t('revoke'),
			variant: 'disabled',
		},
		{
			icon: 'PencilSimple',
			route: NamedRoutes.EditRole,
			title: t('edit'),
			variant: 'disabled',
		},
	];

	return (
		<BaseContainer title={t('title')}>
			<Box p={{ base: 4, md: '128px' }}>
				<Heading fontSize='20px' fontWeight='600' mb={14} data-testid='subtitle'>
					{t('subtitle')}
				</Heading>
				<GridDirectAccess directAccesses={directAccesses} />
			</Box>
		</BaseContainer>
	);
};

export default Roles;
