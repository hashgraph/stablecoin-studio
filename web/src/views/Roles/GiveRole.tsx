import { useTranslation } from 'react-i18next';
import { Box, HStack, Text } from '@chakra-ui/react';
import RoleLayout from './RoleLayout';
import { useForm } from 'react-hook-form';
import SwitchController from '../../components/Form/SwitchController';
import InputNumberController from '../../components/Form/InputNumberController';
import { fakeOptions, fields } from './constans';

const supplier = 'Supplier';

const GiveRole = () => {
	const { t } = useTranslation(['global', 'roles']);
	const {
		control,
		formState: { isValid },
		register,
		watch,
	} = useForm({ mode: 'onChange' });
	register(fields.supplierQuantitySwitch, { value: true });

	const account = watch(fields.account);
	const amount = watch(fields.amount);
	const infinity = watch(fields.supplierQuantitySwitch);
	const role = watch(fields.role);

	const handleSubmit = () => {
		console.log(`Give role ${role.label} to account ${account}`);
		if (role.label === supplier) {
			if (infinity) {
				console.log('With a infinity quantity of tokens');
			} else {
				console.log(`With a supply of ${amount} tokens`);
			}
		}
	};

	const renderSupplierQuantity = () => {
		return (
			<Box data-testid='supplier-quantity'>
				<HStack>
					<Text>{t('roles:giveRole.supplierQuantityQuestion')}</Text>
				</HStack>
				<HStack mt='20px'>
					<Text mr='10px'>{t('roles:giveRole.switchLabel')}</Text>
					<SwitchController control={control} name={fields.supplierQuantitySwitch} />
				</HStack>
				{!infinity && (
					<Box mt='20px'>
						<InputNumberController
							data-testid='input-supplier-quantity'
							rules={{
								required: t('global:validations.required'),
							}}
							isRequired
							control={control}
							name={fields.amount}
							placeholder={t('roles:giveRole.supplierQuantityInputPlaceholder')}
						/>
					</Box>
				)}
			</Box>
		);
	};

	return (
		<RoleLayout
			accountLabel={t('roles:giveRole.accountLabel')}
			accountPlaceholder={t('roles:giveRole.accountPlaceholder')}
			buttonConfirmEnable={isValid}
			control={control}
			onConfirm={() => handleSubmit()}
			options={fakeOptions}
			selectorLabel={t('roles:giveRole.selectLabel')}
			selectorPlaceholder={t('roles:giveRole.selectPlaceholder')}
			title={t('roles:give')}
		>
			{role?.label === supplier && renderSupplierQuantity()}
		</RoleLayout>
	);
};

export default GiveRole;
