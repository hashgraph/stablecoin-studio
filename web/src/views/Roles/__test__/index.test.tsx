import Roles from '../';
import { render } from '../../../test/index';
import translations from '../../../translations/en/roles.json';

describe(`<${Roles.name} />`, () => {
	test('should render correctly', () => {
		const component = render(<Roles />);

		expect(component.asFragment()).toMatchSnapshot();
	});

	test('should has title', () => {
		const component = render(<Roles />);
		const header = component.getByTestId('base-container-heading');

		expect(header).toHaveTextContent(translations.title);
	});

	test('should has subtitle', () => {
		const component = render(<Roles />);
		const subtitle = component.getByTestId('subtitle');

		expect(subtitle).toHaveTextContent(translations.subtitle);
	});

	test('should has direct accesses', () => {
		const component = render(<Roles />);

		const directAccesses = component.getAllByRole('button');
		expect(directAccesses).toHaveLength(3);
	});
});
