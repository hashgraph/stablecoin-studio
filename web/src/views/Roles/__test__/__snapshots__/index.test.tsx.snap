// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Roles /> should render correctly 1`] = `
<DocumentFragment>
  <div
    class="chakra-stack css-16oka5y"
  >
    <h2
      class="chakra-heading css-llrpa2"
      data-testid="base-container-heading"
    >
      Role management
    </h2>
    <div
      class="css-1b6lda7"
      data-testid="base-container-children"
    >
      <div
        class="css-src872"
      >
        <h2
          class="chakra-heading css-1brd3j2"
          data-testid="subtitle"
        >
          Give / Revoke roles
        </h2>
        <div
          class="css-qbodi0"
        >
          <div
            class="css-180e663"
          >
            <button
              class="css-yidqqh"
              data-testid="direct-access-giveRole"
            >
              <div
                class="css-bv4l9q"
              >
                <svg
                  class="chakra-icon css-1f9qa95"
                  data-testid="direct-access-PlusCircle"
                  fill="currentColor"
                  focusable="false"
                  height="1em"
                  viewBox="0 0 256 256"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <rect
                    fill="none"
                    height="256"
                    width="256"
                  />
                  <circle
                    cx="128"
                    cy="128"
                    fill="none"
                    r="96"
                    stroke="currentColor"
                    stroke-miterlimit="10"
                    stroke-width="16"
                  />
                  <line
                    fill="none"
                    stroke="currentColor"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                    stroke-width="16"
                    x1="88"
                    x2="168"
                    y1="128"
                    y2="128"
                  />
                  <line
                    fill="none"
                    stroke="currentColor"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                    stroke-width="16"
                    x1="128"
                    x2="128"
                    y1="88"
                    y2="168"
                  />
                </svg>
              </div>
              <p
                class="chakra-text css-1mbt7nu"
                data-testid="direct-access-Give role"
              >
                Give role
              </p>
            </button>
          </div>
          <div
            class="css-180e663"
          >
            <button
              class="css-1qzs5o8"
              data-testid="direct-access-revokeRole"
              disabled=""
            >
              <div
                class="css-bv4l9q"
              >
                <svg
                  class="chakra-icon css-1f9qa95"
                  data-testid="direct-access-MinusCircle"
                  fill="currentColor"
                  focusable="false"
                  height="1em"
                  viewBox="0 0 256 256"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <rect
                    fill="none"
                    height="256"
                    width="256"
                  />
                  <circle
                    cx="128"
                    cy="128"
                    fill="none"
                    r="96"
                    stroke="currentColor"
                    stroke-miterlimit="10"
                    stroke-width="16"
                  />
                  <line
                    fill="none"
                    stroke="currentColor"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                    stroke-width="16"
                    x1="88"
                    x2="168"
                    y1="128"
                    y2="128"
                  />
                </svg>
              </div>
              <p
                class="chakra-text css-1mbt7nu"
                data-testid="direct-access-Revoke role"
              >
                Revoke role
              </p>
            </button>
          </div>
          <div
            class="css-180e663"
          >
            <button
              class="css-1qzs5o8"
              data-testid="direct-access-editRole"
              disabled=""
            >
              <div
                class="css-bv4l9q"
              >
                <svg
                  class="chakra-icon css-1f9qa95"
                  data-testid="direct-access-PencilSimple"
                  fill="currentColor"
                  focusable="false"
                  height="1em"
                  viewBox="0 0 256 256"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <rect
                    fill="none"
                    height="256"
                    width="256"
                  />
                  <path
                    d="M92.7,216H48a8,8,0,0,1-8-8V163.3a7.9,7.9,0,0,1,2.3-5.6l120-120a8,8,0,0,1,11.4,0l44.6,44.6a8,8,0,0,1,0,11.4l-120,120A7.9,7.9,0,0,1,92.7,216Z"
                    fill="none"
                    stroke="currentColor"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                    stroke-width="16"
                  />
                  <line
                    fill="none"
                    stroke="currentColor"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                    stroke-width="16"
                    x1="136"
                    x2="192"
                    y1="64"
                    y2="120"
                  />
                </svg>
              </div>
              <p
                class="chakra-text css-1mbt7nu"
                data-testid="direct-access-Edit role"
              >
                Edit role
              </p>
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
  <span
    hidden=""
    id="__chakra_env"
  />
</DocumentFragment>
`;
